File: HappySupremacyConfig/Code/Client/HappySupremacy_CombineCardsCFG.cpp
#include "stdafx.h"
#include "tinyxml2/tinyxml2.h"
#include "LuaCfgHelper.h"
#include <iostream>
#include <boost/algorithm/string.hpp>
#include "HappySupremacy_CombineCardsCFG.h"
std::auto_ptr<HappySupremacy_CombineCardsCFG> HappySupremacy_CombineCardsCFG::msSingleton(nullptr);

int HappySupremacy_CombineCardsCFG::GetCount()
{
	return (int)mMapData.size();
}

const HappySupremacy_CombineCardsCFGData* HappySupremacy_CombineCardsCFG::GetData(int CombineCardID)
{
	auto it = mMapData.find(CombineCardID);
	if (it != mMapData.end())
	{
		return &it->second;
	}
	return NULL;
}

const std::map<int, HappySupremacy_CombineCardsCFGData>& HappySupremacy_CombineCardsCFG::GetMapData()
{
	return mMapData;
}

void HappySupremacy_CombineCardsCFG::Load()
{
	tinyxml2::XMLDocument xmlDoc;
	std::string content = FileUtils::getInstance()->getStringFromFile("Config/HappySupremacy_CombineCardsCFG.xml");
	auto result = xmlDoc.Parse(content.c_str(), content.length());
	if (result != tinyxml2::XML_SUCCESS)
	{
		CCLOGERROR("Result:%d", result);
		CCASSERT(false, "result != tinyxml2::XML_SUCCESS");
		return;
	}
	auto root = xmlDoc.RootElement();
	if (root == NULL)
	{
		CCASSERT(false, "root == NULL");
		return;
	}
	auto element = root->FirstChildElement("Data");
	while (element != NULL)
	{
		HappySupremacy_CombineCardsCFGData data;
		data.mCombineCardID = element->IntAttribute("CombineCardID");
		data.mCombineCardsName = element->Attribute("CombineCardsName");
		if (mMapData.find(data.mCombineCardID) != mMapData.end())std::cout <<"data refind:" << data.mCombineCardID << std::endl;
		CCASSERT(mMapData.find(data.mCombineCardID) == mMapData.end(), "data.mCombineCardID is exists");
		mMapData.insert(std::make_pair(data.mCombineCardID, data));
		element = element->NextSiblingElement();
	}
	CCLOG("HappySupremacy_CombineCardsCFG Loaded. Load Data:%u", mMapData.size());
}

void HappySupremacy_CombineCardsCFG::LoadLua()
{
	LuaEngine::getInstance()->executeScriptFile("Config/HappySupremacy_CombineCardsCFG");
	lua_State* L = LuaEngine::getInstance()->getLuaStack()->getLuaState();
	lua_getglobal(L, "HappySupremacy_CombineCardsCFG");
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushstring(L, "datas");
	lua_gettable(L, -2);
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushnil(L);
	while(lua_next(L, 2))
	{
		CCASSERT(lua_istable(L, -1) == 1, "is not table");
		HappySupremacy_CombineCardsCFGData data;
		LuaCfgHelper::readInt(L, "CombineCardID", data.mCombineCardID);
		LuaCfgHelper::readString(L, "CombineCardsName", data.mCombineCardsName);
		if (mMapData.find(data.mCombineCardID) != mMapData.end())std::cout <<"data refind:" << data.mCombineCardID << std::endl;
		CCASSERT(mMapData.find(data.mCombineCardID) == mMapData.end(), "data.mCombineCardID is exists");
		mMapData.insert(std::make_pair(data.mCombineCardID, data));
		lua_pop(L, 1);
	}
	lua_settop(L, 0);
	CCLOG("HappySupremacy_CombineCardsCFG Loaded. Load Data:%u", mMapData.size());
}

void HappySupremacy_CombineCardsCFG::Reload()
{
	mMapData.clear();
	Load();
}

HappySupremacy_CombineCardsCFG* HappySupremacy_CombineCardsCFG::GetSingleton()
{
	if (msSingleton.get() == nullptr)
	{
		msSingleton.reset(new HappySupremacy_CombineCardsCFG());
	}
	return msSingleton.get();
}

File: HappySupremacyConfig/Lua/HappySupremacy_CombineCardsCFG.lua
HappySupremacy_CombineCardsCFG = {}

function HappySupremacy_CombineCardsCFG:getData(key)
	if self.datas == nil then
		return nil
	end
	return self.datas[key]
end

function HappySupremacy_CombineCardsCFG:init()
	self.datas = {}
	self.datas[1] = {CombineCardID = 1, CombineCardsName = "至尊宝"}
	self.datas[2] = {CombineCardID = 2, CombineCardsName = "双天"}
	self.datas[3] = {CombineCardID = 3, CombineCardsName = "双地"}
	self.datas[4] = {CombineCardID = 4, CombineCardsName = "双人"}
	self.datas[5] = {CombineCardID = 5, CombineCardsName = "双和"}
	self.datas[6] = {CombineCardID = 6, CombineCardsName = "双梅"}
	self.datas[7] = {CombineCardID = 7, CombineCardsName = "双长"}
	self.datas[8] = {CombineCardID = 8, CombineCardsName = "双板凳"}
	self.datas[9] = {CombineCardID = 9, CombineCardsName = "双斧头"}
	self.datas[10] = {CombineCardID = 10, CombineCardsName = "双红头"}
	self.datas[11] = {CombineCardID = 11, CombineCardsName = "双铜棰"}
	self.datas[12] = {CombineCardID = 12, CombineCardsName = "双幺五"}
	self.datas[13] = {CombineCardID = 13, CombineCardsName = "杂九"}
	self.datas[14] = {CombineCardID = 14, CombineCardsName = "杂八"}
	self.datas[15] = {CombineCardID = 15, CombineCardsName = "杂七"}
	self.datas[16] = {CombineCardID = 16, CombineCardsName = "杂五"}
	self.datas[17] = {CombineCardID = 17, CombineCardsName = "天九王"}
	self.datas[18] = {CombineCardID = 18, CombineCardsName = "天槓"}
	self.datas[19] = {CombineCardID = 19, CombineCardsName = "地槓"}
end

HappySupremacy_CombineCardsCFG:init()
